deps/build.jl

```diff
@@ -1,133 +1,61 @@
-    "libavfilter4-ffmpeg"   => libavfilter,
-    "libavformat55-ffmpeg"  => libavformat,
-    "libavutil52-ffmpeg"    => libavutil,
-    #"libswresample0-ffmpeg" => libswresample,
-    "libswscale2-ffmpeg"    => libswscale,
-    # ffmpeg is again available in standard packages as of ubuntu 15.04
-    "libavcodec-ffmpeg56"   => libavcodec,
-    "libavdevice-ffmpeg56"  => libavdevice,
-    "libavfilter-ffmpeg5"   => libavfilter,
-    "libavformat-ffmpeg56"  => libavformat,
-    "libavutil-ffmpeg54"    => libavutil,
-    "libswscale-ffmpeg3"    => libswscale,
-    #"libavresample-ffmpeg2" => libavresample,
-)
-
-#if have_swresample
-#    apt_packages["libswresample0-ffmpeg"] = libswresample
-#end
-
-#if have_avresample
-#    apt_packages["libavresample1"] = libavresample
-#end
-
-provides(AptGet,
-         apt_packages)
-
-provides(Yum,
-         Dict("ffmpeg" => libav_libs))
-
-provides(Pacman,
-         Dict("ffmpeg" => libav_libs))
-
-provides(Sources,
-         URI("http://www.ffmpeg.org/releases/ffmpeg-4.1.tar.gz"),
-         libav_libs)
-
-provides(BuildProcess, Autotools(configure_options=["--enable-gpl"]), libav_libs, os = :Unix)
-
-@BinDeps.install Dict(
-    :libavcodec => :libavcodec,
-    :libavformat => :libavformat,
-    :libavutil => :libavutil,
-    :libswscale => :libswscale,
-    :libavfilter => :libavfilter,
-    :libavdevice => :libavdevice
+using BinaryProvider
+
+# Parse some basic command-line arguments
+const verbose = "--verbose" in ARGS
+const prefix = Prefix(get([a for a in ARGS if a != "--verbose"], 1, joinpath(@__DIR__, "usr")))
+
+libpath = joinpath(@__DIR__, "usr/bin") #for forcing LibraryProduct to find libs in bin dir (as ffmpeg likes to do)
+# These are the two binary objects we care about
+products = Product[
+    ExecutableProduct(prefix, "ffmpeg", :ffmpeg),
+    ExecutableProduct(prefix, "ffprobe", :ffprobe),
+    ExecutableProduct(prefix, "ffplay", :ffplay),
+
+    LibraryProduct(libpath, [["libavcodec-ffmpeg.so.","libavcodec.","libavcodec.so.","libavcodec.ffmpeg.so.","avcodec-"].*["53" "54" "55" "56" "57" "58"].*[".dylib"]...], :libavcodec),
+    LibraryProduct(libpath, [["libavformat-ffmpeg.so.","libavformat.","libavformat.so.","libavformat.ffmpeg.so.","avformat-"].*["53" "54" "55" "56" "57" "58"]...], :libavformat),
+    LibraryProduct(libpath, [["libavutil-ffmpeg.so.", "libavutil.","libavutil.so.", "libavutil.ffmpeg.so.", "avutil-"].*["51" "52" "54" "55" "56"]...], :libavutil),
+    LibraryProduct(libpath, [["libswscale-ffmpeg.so.","libswscale.","libswscale.so.","libswscale.ffmpeg.so.","swscale-"].*["2" "3" "4" "5"]...], :libswscale),
+    LibraryProduct(libpath, [["libavfilter-ffmpeg.so.","libavfilter.","libavfilter.so.","libavfilter.ffmpeg.so.","avfilter-"].*["2" "3" "4" "5" "6" "7"]...], :libavfilter),
+    LibraryProduct(libpath, [["libavdevice-ffmpeg.so.","libavdevice.","libavdevice.so.","libavdevice.ffmpeg.so.","avdevice-"].*["53" "54" "55" "56" "57" "58"]...], :libavdevice),
+]
```

Comment:
These `LibraryProduct` definitions are probably more complex than they need to be.  Do you actually care that the versions are one of `53:58`, or can it just be any version number?  Ideally, you'd just be giving names like `["libavcodec", "libavcodec.ffmpeg"]` and BinaryProvider will take care of dealing with `.53.so`, or `.so`, or `.dylib`, or whatever.  If there is a definitive list of what these things can be called, I can help you make the simplest possible `LibraryProduct` line.

Reply:
This format/list is inherited from the BinDeps setup, and the whole package is built out for specific library versions given the degree of changes that ffmpeg has had since the earlier supported versions, so I believe we do need to be specific, but perhaps ranges would suffice.

Reply:
Since you have the actual URLs below, I bet we can cut this down to just the versions within the tarballs and make this a lot simpler.

Reply:
The problem is that I haven't yet found linux platform-specific shared binaries, which would be ideal to control version like we're trying in windows and MacOS. 
Given linux typically(?) comes preinstalled with ffmpeg, the coverage of all the lib versions is needed for that. Perhaps there's an alternative that I'm missing..